### PR TITLE
Fix Runtime_76194 test

### LINKED ISSUE #78758: Runtime_76194 fails on ARM32
The test added in #76341 fails on ARM32.

Work items: https://dev.azure.com/dnceng-public/public/_build/results?buildId=90661&view=ms.vss-test-web.build-test-results-tab

```scala
    JIT/Regression/JitBlue/Runtime_76194/Runtime_76194/Runtime_76194.sh [FAIL]
      Unhandled exception. System.NullReferenceException: Object reference not set to an instance of an object.
         at Runtime_76194.Read1(Byte* location)
         at Runtime_76194.Main()
```

cc @EgorBo

### PR DESCRIPTION
**Pull request by EgorBo**
Should fix https://github.com/dotnet/runtime/issues/78758

cc @jakobbotsch

### CODE REVIEW

File: src/tests/JIT/Regression/JitBlue/Runtime_76194/Runtime_76194.cs [hunk 1]
```diff
@@ -127,6 +121,7 @@ public CrossVirtualAlloc()
             if (_ptr != null && mprotect(_ptr + PageSize, PageSize, PROT_NONE) != 0)
             {
                 munmap(_ptr, 2 * PageSize);
+                _ptr = null;
             }
```
Comment (jakobbotsch) [inline src/tests/JIT/Regression/JitBlue/Runtime_76194/Runtime_76194.cs:125]: `mmap` does not return null on failure, so the check needs to be for -1

Comment (jakobbotsch) [reply src/tests/JIT/Regression/JitBlue/Runtime_76194/Runtime_76194.cs:125]: `VirtualFree` above should also get this double free fix

Comment (EgorBo) [reply src/tests/JIT/Regression/JitBlue/Runtime_76194/Runtime_76194.cs:132]: I don't see what's wrong with VirtualFree

**Comment by EgorBo** [comment]
`mmap(null, 2 * PageSize, PROT_READ | PROT_WRITE, MAP_PRIVATE | MAP_ANONYMOUS, -1, 0);`

returns -1 on `Linux_musl arm` with errno 22 which should be EINVAL

```
EINVAL
  We don't like addr, length, or offset (e.g., they are too large, or not aligned on a page boundary).
EINVAL
  (since Linux 2.6.12) length was 0.
EINVAL
  flags contained neither MAP_PRIVATE or MAP_SHARED, or contained both of these values.
```

Log: https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-78760-merge-876e510b87fc47e5bb/JIT.Regression/1/console.affca64c.log?helixlogtype=result

Test: https://github.com/dotnet/runtime/blob/ce5929a75b4f7272dcc1000b16a13e45e0d4d997/src/tests/JIT/Regression/JitBlue/Runtime_76194/Runtime_76194.cs

@jkotas @jakobbotsch does it ring a bell to you by any chance what could be the reason?
I mean in theory the test is expected to reserve 2 pages (in a loop of 100 iterations)

**Comment by jakobbotsch** [comment]
I think the problem might be that the value of `MAP_ANONYMOUS` is different depending on the standard lib. I can see that we define `MAP_ANONYMOUS = 0x10` in our own sources:

https://source.dot.net/#System.IO.MemoryMappedFiles/src/libraries/Common/src/Interop/Unix/System.Native/Interop.MMap.cs,89e5b8f2195492e1

Likely you can pinvoke this SystemNative version and it will handle the conversions for us?

**Comment by jakobbotsch** [comment]
This also explains why we only were seeing this issue in outerloop runs.

**Comment by jkotas** [comment]
> Likely you can pinvoke this SystemNative version and it will handle the conversions for us?

If the tests need native dependencies, we typically create a test local native dependency for the tests that does what the test needs.

File: src/tests/JIT/Regression/JitBlue/Runtime_76194/CrossplatVirtualAlloc.cpp [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+        void* ptr = VirtualAlloc(reservePtr, size, MEM_COMMIT, PAGE_READWRITE);
+        if (ptr == nullptr)
+        {
+            VirtualFree(reservePtr, 0, MEM_RELEASE);
+        }
+        return ptr;
+    }
+#else
+    uint8_t* ptr = (uint8_t*)mmap(nullptr, size * 2, PROT_READ | PROT_WRITE, MAP_PRIVATE | MAP_ANONYMOUS, -1, 0)
+    if (ptr != (void*)-1)
```
Comment (jkotas) [inline src/tests/JIT/Regression/JitBlue/Runtime_76194/CrossplatVirtualAlloc.cpp:27]:
```suggestion
    if (ptr != MAP_FAILED)
```

File: src/tests/JIT/Regression/JitBlue/Runtime_76194/CrossplatVirtualAlloc.cpp [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+        {
+            VirtualFree(reservePtr, 0, MEM_RELEASE);
+        }
+        return ptr;
+    }
+#else
+    uint8_t* ptr = (uint8_t*)mmap(nullptr, size * 2, PROT_READ | PROT_WRITE, MAP_PRIVATE | MAP_ANONYMOUS, -1, 0);
+    if (ptr != (void*)-1)
+    {
+        if (mprotect(ptr + size, size, 0) == 0)
```
Comment (jkotas) [inline src/tests/JIT/Regression/JitBlue/Runtime_76194/CrossplatVirtualAlloc.cpp:29]:
```suggestion
        if (mprotect(ptr + size, size, PROT_NONE) == 0)
```

**Comment by EgorBo** [comment]
@dotnet/jit-contrib @jkotas PTAL, this unblocks outerloop CI

File: src/tests/JIT/Regression/JitBlue/Runtime_76194/Runtime_76194.csproj [hunk 4]
```diff
@@ -3,8 +3,10 @@
     <OutputType>Exe</OutputType>
     <Optimize>True</Optimize>
     <AllowUnsafeBlocks>True</AllowUnsafeBlocks>
+    <CLRTestTargetUnsupported Condition="'$(TargetOS)' == 'OSX'">true</CLRTestTargetUnsupported>
```
Comment (jkotas) [inline src/tests/JIT/Regression/JitBlue/Runtime_76194/Runtime_76194.csproj:6]: What prevents the test from working on macOS?

File: src/tests/JIT/Regression/JitBlue/Runtime_76194/CrossplatVirtualAlloc.cpp [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+
+#include <platformdefines.h>
+#include <stdint.h>
+#ifdef _WIN32
+#include <memoryapi.h>
+#else
+#include <sys/mman.h>
+#endif
+
+extern "C" DLL_EXPORT void* Alloc(uintptr_t size)
```
Comment (jkotas) [inline src/tests/JIT/Regression/JitBlue/Runtime_76194/CrossplatVirtualAlloc.cpp:12]:
```suggestion
extern "C" DLL_EXPORT void* AllocWithGuard(uintptr_t size)
```
This is not plain vanilla Alloc.

**Review by jkotas — APPROVED** [review]

File: src/tests/JIT/Regression/JitBlue/Runtime_76194/Runtime_76194.csproj (hunk 4 above)
Comment (EgorBo) [reply src/tests/JIT/Regression/JitBlue/Runtime_76194/Runtime_76194.csproj:6]: hm.. not sure actually, I think the previous version with DllImport just didn't work on macOS

**Comment by EgorBo** [comment]
Had to disable the tests for some mono targets: Wasm, Android, iOS, FullAOT where other tests with native libs are also ignored.

**Comment by EgorBo** [comment]
Failures are https://github.com/dotnet/runtime/issues/79170 and https://github.com/dotnet/runtime/issues/79255

**Pull request merged**